### PR TITLE
Additional Rosseland opacity fixes

### PR DESCRIPTION
**Pull request by brryan**
I was still running into `infinity` values from the Rosseland mean opacity so I modified the code to only compute the integrand if `alpha > singularity_opac::robust::SMALL()` for that support point. 

Along the way I cleaned up the code for the trapezoidal rule, reducing redundant opacity evaluations. 

Mean opacities can also now be evaluated either with a default per-rho, T, Ye frequency range, or with a common frequency range specified by an overloaded constructor

### CODE REVIEW

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 1]
```diff
@@ -46,6 +46,24 @@ class MeanOpacity {
...
+              Real lNuMax, const int NNu, Real *lambda = nullptr) {
+    MeanOpacityImpl<Opacity, false>(opac, lRhoMin, lRhoMax, NRho, lTMin, lTMax, NT, YeMin, YeMax,
+                NYe, lNuMin, lNuMax, NNu, lambda);
+  }
+
+  template <typename Opacity, bool AUTOFREQ>
+  MeanOpacityImpl(const Opacity &opac, const Real lRhoMin, const Real lRhoMax,
+              const int NRho, const Real lTMin, const Real lTMax, const int NT,
+              const Real YeMin, const Real YeMax, const int NYe, Real lNuMin,
+              Real lNuMax, const int NNu, Real *lambda = nullptr) {
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Does this actually work? I think your constructor **has** to be named the same as the class. You could, however, use tag dispatch or something?

File: singularity-opac/neutrinos/mean_s_opacity_neutrinos.hpp [hunk 2]
```diff
@@ -46,6 +46,24 @@ class MeanSOpacity {
...
+               Real lNuMax, const int NNu, Real *lambda = nullptr) {
+    MeanSOpacityImpl<SOpacity, false>(s_opac, lRhoMin, lRhoMax, NRho, lTMin, lTMax, NT, YeMin, YeMax,
+                 NYe, lNuMin, lNuMax, 100, lambda);
+  }
+
+  template <typename SOpacity, bool AUTOFREQ>
+  MeanSOpacityImpl(const SOpacity &s_opac, const Real lRhoMin, const Real lRhoMax,
+               const int NRho, const Real lTMin, const Real lTMax, const int NT,
+               const Real YeMin, const Real YeMax, const int NYe, Real lNuMin,
+               Real lNuMax, const int NNu, Real *lambda = nullptr) {
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_s_opacity_neutrinos.hpp]: See above concern about impl constructor

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 1 above)
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Does this actually work? I think the constructor has to be named exactly the same as the class. Maybe you need to go to tag dispatch or something?

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 3]
```diff
@@ -68,68 +86,45 @@ class MeanOpacity {
...
-            lkappaPlanck_(iRho, iT, iYe, idx) = lkappaPlanck;
-            lkappaRosseland_(iRho, iT, iYe, idx) = lkappaRosseland;
+            Real kappaPlanck = singularity_opac::robust::ratio(
+                kappaPlanckNum, kappaPlanckDenom);
+            Real kappaRosseland =
+                kappaPlanck > singularity_opac::robust::SMALL()
+                    ? singularity_opac::robust::ratio(kappaRosselandDenom,
+                                                      kappaRosselandNum)
+                    : 0.;
+            printf("kappas: %e %e\n", kappaPlanck, kappaRosseland);
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Should this printf still be here? If so might be good to make it more descriptive

File: singularity-opac/photons/mean_opacity_photons.hpp [hunk 4]
```diff
@@ -45,6 +45,23 @@ class MeanOpacity {
...
+              Real lNuMin, Real lNuMax, const int NNu, Real *lambda = nullptr) {
+    MeanOpacityImpl<Opacity, false>(opac, lRhoMin, lRhoMax, NRho, lTMin, lTMax,
+                                    NT, lNuMin, lNuMax, NNu, lambda);
+  }
+
+  template <typename Opacity, bool AUTOFREQ>
+  MeanOpacityImpl(const Opacity &opac, const Real lRhoMin, const Real lRhoMax,
+                  const int NRho, const Real lTMin, const Real lTMax,
+                  const int NT, Real lNuMin, Real lNuMax, const int NNu,
+                  Real *lambda = nullptr) {
```
Comment (Yurlungur) [inline singularity-opac/photons/mean_opacity_photons.hpp]: See above concern about `impl`

File: singularity-opac/photons/mean_s_opacity_photons.hpp [hunk 5]
```diff
@@ -46,6 +46,24 @@ class MeanSOpacity {
...
+               Real *lambda = nullptr) {
+    MeanSOpacityImpl<SOpacity, false>(s_opac, lRhoMin, lRhoMax, NRho, lTMin,
+                                      lTMax, NT, lNuMin, lNuMax, NNu, lambda);
+  }
+
+  template <typename SOpacity, bool AUTOFREQ>
+  MeanSOpacityImpl(const SOpacity &s_opac, const Real lRhoMin,
+                   const Real lRhoMax, const int NRho, const Real lTMin,
+                   const Real lTMax, const int NT, Real lNuMin, Real lNuMax,
+                   const int NNu, Real *lambda = nullptr) {
```
Comment (Yurlungur) [inline singularity-opac/photons/mean_s_opacity_photons.hpp]: see above question about `impl` constructor

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Overall nice cleanup. :+1: Looks like tests are failing though and I suspect I know why. See below:

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 1 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Oh yeah you're right, I meant `MeanOpacityImpl` to be a `void` function. I changed that here and elsewhere and the tests pass without memory failures. I have no idea what the compiler was doing before that change...

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 3 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Oops yeah that was garbage, gone

**Comment by brryan** [comment]
OK this should be good now @Yurlungur

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 6]
```diff
@@ -46,6 +46,27 @@ class MeanOpacity {
...
+              const int NRho, const Real lTMin, const Real lTMax, const int NT,
+              const Real YeMin, const Real YeMax, const int NYe, Real lNuMin,
+              Real lNuMax, const int NNu, Real *lambda = nullptr) {
+    MeanOpacityImpl<Opacity, false>(opac, lRhoMin, lRhoMax, NRho, lTMin, lTMax,
+                                    NT, YeMin, YeMax, NYe, lNuMin, lNuMax, NNu,
+                                    lambda);
+  }
+
+  template <typename Opacity, bool AUTOFREQ>
+  void MeanOpacityImpl(const Opacity &opac, const Real lRhoMin,
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: As a void function should `MeanOpacityImpl` have public scope? Also if private scope, should have trailing underscore

**Review by Yurlungur — CHANGES_REQUESTED** [review]

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 6 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Ah yes these four `void` functions should be `private` and have a trailing underscore -- done, thanks

**Review by Yurlungur — APPROVED** [review]
:+1:

**Pull request merged**